src/gui/src/staGui.cpp

```diff
@@ -785,4 +909,437 @@ void TimingPathRenderer::highlightStage(gui::Painter& painter,
+  const auto text_anchor = gui::Painter::Anchor::CENTER;
+  const double text_margin = 2.0;
+  painter.setPen(gui::Painter::white, true);
+  for (const auto& [level, pins] : map_) {
+    for (const auto& pin : pins) {
+      const odb::Rect pin_rect = pin->getPinLargestBox();
+      const odb::Point pin_pt(0.5 * (pin_rect.xMin() + pin_rect.xMax()),
+                              0.5 * (pin_rect.yMin() + pin_rect.yMax()));
+
+      const std::string text = std::to_string(level);
+      const odb::Rect text_bound = painter.stringBoundaries(pin_pt.x(), pin_pt.y(), text_anchor, text);
+
+      if (text_bound.dx() < text_margin * pin_rect.dx() && text_bound.dy() < text_margin * pin_rect.dy()) {
+        painter.drawString(pin_pt.x(), pin_pt.y(), text_anchor, text);
+      }
+    }
+  }
+
+  // draw legend
+  const int legend_keys = 5;
+  const int color_count = color_generator_.getColorCount();
+  auto* units = sta_->units()->timeUnit();
+  const std::string text_units = std::string(units->scaleAbreviation()) + units->suffix();
+  std::vector<std::pair<int, std::string>> legend;
+  for (int i = 0; i < legend_keys; i++) {
+    const double scale = static_cast<double>(i) / (legend_keys - 1);
+    const int color_index = color_count * scale;
+    const double slack = max_timing_ - timing_range * scale;
+    const std::string text = units->asString(slack) + text_units;
+    legend.push_back({color_index, text});
+  }
+  std::reverse(legend.begin(), legend.end());
+  color_generator_.drawLegend(painter, legend);
+}
+
+void TimingConeRenderer::getFaninCone(sta::Pin* source_pin, DepthMapSet& depth_map)
+{
+  sta::PinSeq pins_to;
+  pins_to.push_back(source_pin);
+
+  auto* pins = sta_->findFaninPins(&pins_to,
+                                   true, // flat
+                                   false, // startpoints_only
+                                   0,
+                                   0,
+                                   true, // thru_disabled
+                                   true); // thru_constants
+
+  sta::SearchPred2 srch_pred(sta_);
+  sta::BfsBkwdIterator bfs(sta::BfsIndex::other, &srch_pred, sta_);
+
+  DepthMapSet depth_map_set;
+  getCone(source_pin, pins, bfs, depth_map_set);
+  for (auto& [level, pin_list] : depth_map_set) {
+    depth_map[-level].insert(pin_list.begin(), pin_list.end());
+  }
+}
+
+void TimingConeRenderer::getFanoutCone(sta::Pin* source_pin, DepthMapSet& depth_map)
+{
+  sta::PinSeq pins_from;
+  pins_from.push_back(source_pin);
+
+  auto* pins = sta_->findFanoutPins(&pins_from,
+                                    true, // flat
+                                    false, // startpoints_only
+                                    0,
+                                    0,
+                                    true, // thru_disabled
+                                    true); // thru_constants
+  sta::SearchPred2 srch_pred(sta_);
+
+  sta::BfsFwdIterator bfs(sta::BfsIndex::other, &srch_pred, sta_);
```

Comment:
This seems rather bizarre.
findFanoutPins tells you the pin the fanout cone (strangely named "timing cone") so what is the point of doing a BFS after that? Just to find the logic levels wrt to the source pin? Why? Sta::findFanoutPins is over burdened with the symantics of sdc get_fanout options so it has to do a DFS. Just using BfsIterator makes more sense.

Reply:
removed the fanin and fanout calls

Reply:
I ended up removing the BFS and just using the output of fanout and fanin pins, since the levels in the BFS and the distance from the "source pin" didn't match up.